File: src/lib/permissions.ts
// Module-level access ("can this role touch Employees/Onboarding/Attendance
// at all") is governed by the configurable permission matrix a Super Admin
// edits on the User Management page — see backend/app/services/permissions.py
// and PermissionMatrixEditor (@/components/permission-matrix.tsx). It is NOT
// a hardcoded role list on the frontend anymore: every check below reads the
// signed-in account's own `permissions` array (populated by GET /auth/me),
// so a matrix edit takes effect immediately without a frontend redeploy.
// Hiding a button/nav item/page here is a UX nicety, not the real access
// control — the backend enforces the same permission on every route
// regardless of what the UI shows (see app/core/auth.py's require_permission).
//
// Two things stay OUT of the matrix, fixed in code on both ends:
//   - Admin and Super Admin bypass the matrix entirely — always full access
//     everywhere except editing the matrix itself, which is Super Admin only.
//     See FULL_ACCESS_ROLES/isFullAccessRole below.
//   - The SOP-mandated *field-level* splits inside Onboarding (Recruitment
//     Lead vs Onboarding Specialist's checklist columns) and Attendance
//     (Projects prepares, only HR approves/sends) — these come from written
//     SOPs, not a configurable preference, so they stay role-based via
//     canEditOnboardingField below and the backend's ROLE_FIELD_ACCESS /
//     require_violation_approver.

import type { AccountProfile, AccountRole, Permission } from "@/lib/session";

// Mirrors FULL_ACCESS_ROLES in backend/app/services/permissions.py.
export const FULL_ACCESS_ROLES: ReadonlySet<AccountRole> = new Set(["admin", "super_admin"]);

export function isFullAccessRole(role: AccountRole | undefined): boolean {
  return !!role && FULL_ACCESS_ROLES.has(role);
}

// Mirrors PERMISSION_LABELS in backend/app/services/permissions.py — kept as
// a frontend copy rather than fetched, since the only endpoint that returns
// these (GET /permissions/matrix) is Super-Admin-only and the Profile page's
// "My Permissions" section has to work for every role, reading nothing but
// the already-fetched account.permissions from GET /auth/me.
export const PERMISSION_LABELS: Record<Permission, { title: string; description: string }> = {
  "employees.view": {
    title: "View Employees",
    description: "See the Employee Directory.",
  },
  "employees.manage": {
    title: "Manage Employees",
    description: "Create, edit, delete and import employee records.",
  },
  "milestones.view": {
    title: "View Milestones",
    description: "See the Anniversaries and Birthdays pages.",
  },
  "onboarding.view": {
    title: "View Onboarding",
    description: "See the onboarding checklist tracker.",
  },
  "onboarding.manage": {
    title: "Manage Onboarding",
    description: "Add, edit and delete new-hire rows.",
  },
  "attendance.view": {
    title: "View Attendance Violations",
    description: "See the attendance violation tracker.",
  },
  "attendance.manage": {
    title: "Manage Attendance Violations",
    description: "Create, edit, prepare and import violation records.",
  },
  "attendance.approve": {
    title: "Approve Attendance Violations",
    description: "Approve, hold, and send violation emails.",
  },
  "awards.view": {
    title: "View Recognition & Awards",
    description: "See the awards given to employees.",
  },
  "awards.manage": {
    title: "Manage Recognition & Awards",
    description: "Give a new award, and edit or delete existing awards.",
  },
  "benefits.view": {
    title: "View Employee Benefits",
    description: "See the Employee Benefits pages (currently just HMO Management).",
  },
  "benefits.manage": {
    title: "Manage Employee Benefits",
    description:
      "Add, edit and delete benefits records — placeholder until the module is built out.",
  },
};

// Narrower than isFullAccessRole above — a few things (editing the
// permission matrix, triaging the Feedback board's status/priority/reporter
// identity) are Super Admin-only, deliberately excluding a plain Admin. Feed
// this getEffectiveRole(account)'s result, same as isFullAccessRole.
export function isSuperAdminRole(role: AccountRole | undefined): boolean {
  return role === "super_admin";
}

// Mirrors backend/app/core/auth.py's get_effective_role: the role every
// admin/super-admin gate and nav/page permission check should actually use —
// the real persisted role, unless a Super Admin has an active sandbox
// override (see useEnterSandbox/useExitSandbox in @/lib/session). Use this
// instead of reading account.role directly anywhere access is being decided;
// `role` itself stays meaningful only for "who am I really" contexts like the
// Profile page badge and the sandbox control's own visibility.
export function getEffectiveRole(
  account: Pick<AccountProfile, "role" | "sandbox_role"> | null | undefined,
): AccountRole | undefined {
  return account?.sandbox_role ?? account?.role;
}

export function hasPermission(
  permissions: Permission[] | undefined,
  permission: Permission,
): boolean {
  return !!permissions?.includes(permission);
}

export function canViewEmployees(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "employees.view");
}

export function canManageEmployees(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "employees.manage");
}

// Anniversaries + Birthdays — see Permission.MILESTONES_VIEW's own comment
// in backend/app/models/permission.py. Both pages also read from the
// Employee Directory data itself (GET /employees), so a role realistically
// needs employees.view too for the pages to show anything; nav gating in
// @/components/app-sidebar checks both, not just this one.
export function canViewMilestones(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "milestones.view");
}

export function canViewOnboarding(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "onboarding.view");
}

export function canManageOnboarding(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "onboarding.manage");
}

export function canViewAttendance(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "attendance.view");
}

export function canManageAttendance(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "attendance.manage");
}

export function canApproveAttendance(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "attendance.approve");
}

// Recognition & Awards — same "read this module's data" idea as Milestones
// above, but this module has its own real writes too, so it gets a
// canManage alongside canView (see Permission.AWARDS_MANAGE's own comment
// in backend/app/models/permission.py).
export function canViewAwards(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "awards.view");
}

export function canManageAwards(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "awards.manage");
}

// Employee Benefits — placeholder module (HMO Management to start; more
// pages to follow once scoped). Not granted to any role by default; a Super
// Admin turns it on from User Management's permission matrix.
export function canViewBenefits(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "benefits.view");
}

export function canManageBenefits(permissions: Permission[] | undefined): boolean {
  return hasPermission(permissions, "benefits.manage");
}

// Mirrors ROLE_FIELD_ACCESS in backend/app/api/routes/new_hires.py — the New
// Hire Onboarding Tracker SOP's protected-range split: Recruitment Lead owns
// items 1-2, Onboarding Specialist owns items 4-7, item 3 (Welcome Email
// Sent) is shared since it depends on whichever person is available first.
// Fixed by the SOP, not matrix-configurable — see the module comment above.
// Used to disable individual checklist checkboxes per role in
// src/routes/onboarding.tsx — hiding/disabling is a UX nicety here too; the
// backend enforces the same matrix on every PATCH regardless of what the UI
// allows clicking.
const ONBOARDING_FIELD_ACCESS: Record<string, ReadonlySet<AccountRole>> = {
  joDiscussion: new Set(["admin", "super_admin", "recruitment_lead"]),
  confirmationSigned: new Set(["admin", "super_admin", "recruitment_lead"]),
  welcomeEmailSent: new Set(["admin", "super_admin", "recruitment_lead", "onboarding_specialist"]),
  newHireInfo: new Set(["admin", "super_admin", "onboarding_specialist"]),
  idPhoto: new Set(["admin", "super_admin", "onboarding_specialist"]),
  credentialsCreated: new Set(["admin", "super_admin", "onboarding_specialist"]),
  onboardingDay: new Set(["admin", "super_admin", "onboarding_specialist"]),
};

export function canEditOnboardingField(
  role: AccountRole | undefined,
  field: keyof typeof ONBOARDING_FIELD_ACCESS,
): boolean {
  return !!role && (ONBOARDING_FIELD_ACCESS[field]?.has(role) ?? false);
}

File: backend/app/services/permissions.py
"""Reads and writes the permission matrix — see app/models/permission.py for
what this does and doesn't govern.
"""

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.account import Account, AccountRole
from app.models.activity_log import ActivityCategory
from app.models.permission import Permission, RolePermission

# Bypass the matrix entirely — always every permission, matrix rows or not.
FULL_ACCESS_ROLES = {AccountRole.ADMIN, AccountRole.SUPER_ADMIN}

# Every "view" permission — what a restricted account (Account.is_restricted)
# keeps regardless of role; every "manage"/"approve" permission is stripped.
VIEW_PERMISSIONS: frozenset[Permission] = frozenset(
    {
        Permission.EMPLOYEES_VIEW,
        Permission.MILESTONES_VIEW,
        Permission.ONBOARDING_VIEW,
        Permission.ATTENDANCE_VIEW,
        Permission.AWARDS_VIEW,
    }
)

# Which permission governs seeing each Activity Log category — None means
# "admin/super_admin only, not matrix-configurable" (Access/Data/System are
# security- and infrastructure-level, not a specific module a non-admin role
# owns). See visible_activity_categories below.
CATEGORY_PERMISSION: dict[ActivityCategory, Permission | None] = {
    ActivityCategory.EMPLOYEE: Permission.EMPLOYEES_VIEW,
    ActivityCategory.ONBOARDING: Permission.ONBOARDING_VIEW,
    ActivityCategory.ATTENDANCE: Permission.ATTENDANCE_VIEW,
    ActivityCategory.ACCESS: None,
    ActivityCategory.DATA: None,
    ActivityCategory.SYSTEM: None,
}

# The six roles a Super Admin actually configures. Admin/Super Admin are
# deliberately excluded — they're never anything but full-access, so a row
# for them in the matrix would just be confusing dead UI.
MATRIX_ROLES: list[AccountRole] = [
    AccountRole.PEOPLE_OPS,
    AccountRole.HR,
    AccountRole.PROJECTS,
    AccountRole.RECRUITMENT_LEAD,
    AccountRole.ONBOARDING_SPECIALIST,
    AccountRole.VIEWER,
]

PERMISSION_LABELS: dict[Permission, dict[str, str]] = {
    Permission.EMPLOYEES_VIEW: {
        "title": "View Employees",
        "description": "See the Employee Directory.",
    },
    Permission.EMPLOYEES_MANAGE: {
        "title": "Manage Employees",
        "description": "Create, edit, delete and import employee records.",
    },
    Permission.MILESTONES_VIEW: {
        "title": "View Milestones",
        "description": "See the Anniversaries and Birthdays pages.",
    },
    Permission.ONBOARDING_VIEW: {
        "title": "View Onboarding",
        "description": "See the onboarding checklist tracker.",
    },
    Permission.ONBOARDING_MANAGE: {
        "title": "Manage Onboarding",
        "description": "Add, edit and delete new-hire rows (still subject to the SOP's "
        "Recruitment Lead / Onboarding Specialist checklist-field split).",
    },
    Permission.ATTENDANCE_VIEW: {
        "title": "View Attendance Violations",
        "description": "See the attendance violation tracker.",
    },
    Permission.ATTENDANCE_MANAGE: {
        "title": "Manage Attendance Violations",
        "description": "Create, edit, prepare and import violation records.",
    },
    Permission.ATTENDANCE_APPROVE: {
        "title": "Approve Attendance Violations",
        "description": "Approve, hold, and send violation emails.",
    },
    Permission.AWARDS_VIEW: {
        "title": "View Recognition & Awards",
        "description": "See the awards given to employees.",
    },
    Permission.AWARDS_MANAGE: {
        "title": "Manage Recognition & Awards",
        "description": "Give a new award to an active employee, and edit or delete existing awards.",
    },
}

# Default grants — matches this app's behavior from *before* the matrix
# existed, so turning the matrix on doesn't silently change anyone's access
# until a Super Admin actually edits it. The migration that creates
# role_permissions seeds exactly this (a later migration adds
# MILESTONES_VIEW to every one of these, matching Milestones' own
# previously-ungated "open to every signed-in role" behavior).
DEFAULT_GRANTS: dict[AccountRole, set[Permission]] = {
    AccountRole.PEOPLE_OPS: {
        Permission.EMPLOYEES_VIEW,
        Permission.EMPLOYEES_MANAGE,
        Permission.MILESTONES_VIEW,
        Permission.AWARDS_VIEW,
        Permission.AWARDS_MANAGE,
    },
    AccountRole.HR: {
        Permission.EMPLOYEES_VIEW,
        Permission.MILESTONES_VIEW,
        Permission.ATTENDANCE_VIEW,
        Permission.ATTENDANCE_MANAGE,
        Permission.ATTENDANCE_APPROVE,
        Permission.AWARDS_VIEW,
    },
    AccountRole.PROJECTS: {
        Permission.EMPLOYEES_VIEW,
        Permission.MILESTONES_VIEW,
        Permission.ATTENDANCE_VIEW,
        Permission.ATTENDANCE_MANAGE,
        Permission.AWARDS_VIEW,
    },
    AccountRole.RECRUITMENT_LEAD: {
        Permission.EMPLOYEES_VIEW,
        Permission.MILESTONES_VIEW,
        Permission.ONBOARDING_VIEW,
        Permission.ONBOARDING_MANAGE,
        Permission.AWARDS_VIEW,
    },
    AccountRole.ONBOARDING_SPECIALIST: {
        Permission.EMPLOYEES_VIEW,
        Permission.MILESTONES_VIEW,
        Permission.ONBOARDING_VIEW,
        Permission.ONBOARDING_MANAGE,
        Permission.AWARDS_VIEW,
    },
    AccountRole.VIEWER: {
        Permission.EMPLOYEES_VIEW,
        Permission.MILESTONES_VIEW,
        Permission.AWARDS_VIEW,
    },
}


async def get_account_permissions(db: AsyncSession, account: Account, *, role: AccountRole) -> set[Permission]:
    """Every permission this account currently has, computed against `role` —
    always pass the *effective* role for this request (see
    app/core/auth.py's get_effective_role), which is account.role unless a
    Super Admin has an active sandbox override. Used both by has_permission()
    below and to populate AccountRead.permissions for the frontend's own
    nav/page gating.

    A restricted account (account.is_restricted) is capped to view-only here
    regardless of what `role` would otherwise grant — including a Super
    Admin's own FULL_ACCESS_ROLES bypass, so restriction really does mean
    "view only everywhere."""
    if role in FULL_ACCESS_ROLES:
        granted = set(Permission)
    else:
        result = await db.execute(select(RolePermission.permission).where(RolePermission.role == role))
        granted = set(result.scalars().all())
    if account.is_restricted:
        granted &= VIEW_PERMISSIONS
    return granted


async def has_permission(db: AsyncSession, account: Account, permission: Permission, *, role: AccountRole) -> bool:
    return permission in await get_account_permissions(db, account, role=role)


async def visible_activity_categories(
    db: AsyncSession, account: Account, *, role: AccountRole
) -> set[ActivityCategory]:
    """Which Activity Log categories this account may currently see — mirrors
    the module-siloed rule everywhere else in the app: Admin/Super Admin
    (unrestricted) see every category, including the admin-only Access/Data/
    System ones; every other account (and any restricted account, full-
    access or not) only sees the categories whose CATEGORY_PERMISSION they
    currently hold."""
    permissions = await get_account_permissions(db, account, role=role)
    if role in FULL_ACCESS_ROLES and not account.is_restricted:
        return set(ActivityCategory)
    return {
        category
        for category, required in CATEGORY_PERMISSION.items()
        if required is not None and required in permissions
    }


async def get_full_matrix(db: AsyncSession) -> dict[AccountRole, set[Permission]]:
    result = await db.execute(select(RolePermission).where(RolePermission.role.in_(MATRIX_ROLES)))
    matrix: dict[AccountRole, set[Permission]] = {role: set() for role in MATRIX_ROLES}
    for row in result.scalars().all():
        matrix[row.role].add(row.permission)
    return matrix


async def replace_matrix(db: AsyncSession, matrix: dict[AccountRole, set[Permission]]) -> None:
    """Wholesale replace: deletes every existing grant for the roles present
    in `matrix` and re-inserts exactly what's given. The Save button sends
    the entire editable matrix at once, so this is one atomic operation
    rather than N per-cell races. Silently ignores any role outside
    MATRIX_ROLES (e.g. if a caller accidentally includes admin/super_admin)
    rather than erroring — those roles just have nothing to store."""
    roles = [role for role in matrix if role in MATRIX_ROLES]
    if not roles:
        return
    await db.execute(delete(RolePermission).where(RolePermission.role.in_(roles)))
    for role in roles:
        for permission in matrix[role]:
            db.add(RolePermission(role=role, permission=permission))
    await db.flush()
